components/common/src/main/java/com/hotels/styx/common/Result.java

```diff
@@ -0,0 +1,98 @@
+/*
+  Copyright (C) 2013-2018 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.common;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents the outcome of attempting to do something.
+ * On a success a value is stored (defaults to "success" if not specified).
+ * On a failure an exception is stored (defaults to a new Exception if not specified).
+ *
+ * @param <T> type of success value
+ */
+public class Result<T> {
```

Comment:
I haven't checked the code that uses this class or thought about it considering we are doing Java, but I believe we wanted to implement this as an Either/Try monad?  So I would expect : map, flatMap...  Scala's try monad is nicer (as in simpler) https://mauricio.github.io/2014/02/17/scala-either-try-and-the-m-word.html , but  it seems there are a few implementations in Java such as  https://github.com/jasongoodwin/better-java-monads/blob/master/src/main/java/com/jasongoodwin/monads/Try.java

Reply:
The goal of this class was simply to add a representation of success or failure that can be stored as a value.

I can see a similarity with the `Try.java` linked, but I'm not sure it is worth adding a new dependency just for this class.

In terms of API, a difference I see is that the `Result` class can be instantiated directly with a call to `success` or `failure` without needing to wrap a block of code, so it is in that sense more decoupled from the try-catch structure and can be used in situations where try-catch does not appear (e.g. using result values in unit tests).